pkg/machine/machineset.go

```diff
@@ -0,0 +1,378 @@
+
+	builder := MachineSetBuilder{
+		apiClient: apiClient,
+		Definition: &machinev1beta1.MachineSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		},
+	}
+
+	if name == "" {
+		builder.errorMsg = "MachineSet 'name' cannot be empty"
+	}
+
+	if namespace == "" {
+		builder.errorMsg = "MachineSet 'namespace' cannot be empty"
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("MachineSet object %s doesn't exist in namespace %s", name, namespace)
+	}
+
+	builder.Definition = builder.Object
+
+	return &builder, nil
+}
+
+// CreateNewWorkerMachineSetFromCopy returns a MachineSet object which was copied from an exiting one on cluster.
+func CreateNewWorkerMachineSetFromCopy(
+	apiClient *clients.Settings,
+	namespace,
+	publicCloud,
+	instanceType string,
+	replicas int32) (*machinev1beta1.MachineSet, error) {
+	// currently only supporting AWS MachineSets
+	if publicCloud != "aws" {
+		return nil, fmt.Errorf("only AWS public cloud is currently supported for copied MachineSet")
+	}
+
+	workerMachineSetList, err := GetWorkerMachineSetList(apiClient, namespace)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not get worker MachineSetList: %w", err)
+	}
+
+	// picking the first worker MachineSet in list
+	baseMs := workerMachineSetList.Items[0]
+
+	glog.V(100).Infof("Creating new MachineSet copy of first existing worker MachineSet: %v",
+		baseMs.Name)
+
+	copiedMachineSet := &machinev1beta1.MachineSet{
+		ObjectMeta: *baseMs.ObjectMeta.DeepCopy(),
+		Spec:       *baseMs.Spec.DeepCopy(),
+	}
+
+	glog.V(100).Infof("Renaming copied MachineSet to: %s", copiedMachineSet.ObjectMeta.Name)
+
+	copiedMachineSet.ObjectMeta.Name = fmt.Sprintf("%v-%v", copiedMachineSet.Name,
+		strings.ReplaceAll(instanceType, ".", "-"))
+
+	glog.V(100).Infof("Updating copied MachineSet name in metadata, selector and template parameter ...")
+
+	copiedMachineSet.ObjectMeta.UID = ""
+	copiedMachineSet.ObjectMeta.ResourceVersion = ""
+
+	// change spec labels
+	copiedMachineSet.Spec.Selector.MatchLabels["machine.openshift.io/cluster-api-machineset"] =
+		copiedMachineSet.ObjectMeta.Name
+	copiedMachineSet.Spec.Template.ObjectMeta.Labels["machine.openshift.io/cluster-api-machineset"] =
+		copiedMachineSet.ObjectMeta.Name
+
+	glog.V(100).Infof("Updating copied MachineSet provider instanceType to: %s", instanceType)
+
+	// currently only AWS public cloud supported when copying MachineSets
+	err = ChangeAWSProviderInstanceType(copiedMachineSet, instanceType)
+	if err != nil {
+		return nil, fmt.Errorf("could not change the provider instanceType in copied MachineSet: %w", err)
+	}
+
+	glog.V(100).Infof("Updating copied MachineSet replicas value to: %v", replicas)
+	copiedMachineSet.Spec.Replicas = &replicas
+
+	return copiedMachineSet, nil
+}
+
+// Get returns MachineSet object if found.
+func (builder *MachineSetBuilder) Get() (*machinev1beta1.MachineSet, error) {
```

Comment:
You don't need to add this method. Since the versioned interface already supplies a Get method (`builder.apiClient.MachineSets(builder.Definition.Namespace).Get`) you can just use it

Reply:
Done